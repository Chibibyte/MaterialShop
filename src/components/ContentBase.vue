<template>
  <v-card>
    <v-container class="white pb-0">
      <v-form
        :disabled="disabled"
        ref="form"
        :readonly="readonly"
        @input="() => validate(formValidate())"
        class="white"
      >
        <SelectOrCreateContent
          :dense="dense"
          v-for="field in contentMeta"
          :key="field.name+field.label"
          v-model="content[field.name]"
          :label="field.label"
        />
      </v-form>
      <slot />
    </v-container>
  </v-card>
</template>

<script>
import SelectOrCreateContent from "@/components/SelectOrCreateContent.vue";
function createContentField(name, label) {
  return { name, label };
}
export default {
  name: "ContentBase",
  components: {
    SelectOrCreateContent
  },
  props: {
    contentMeta: Array, // createContentField
    validate: {
      type: Function,
      default() {}
    },
    value: {
      type: Object,
      default: () => {
        return {};
      }
    },
    disabled: {
      type: Boolean,
      default: false
    },
    readonly: {
      type: Boolean,
      default: false
    },
    dense: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      content: { ...this.$props.value }
    };
  },
  methods: {
    formValidate() {
      return this.$refs.form ? this.$refs.form.validate() : false;
    },
    createMe(...namesAndLables) {
      let contentMeta = [];
      namesAndLables.forEach(nl =>
        contentMeta.push(createContentField(nl[0], nl[1]))
      );
      return {
        props: {
          value: Object,
          validate: {
            type: Function,
            default() {}
          },
          disabled: {
            type: Boolean,
            default: false
          },
          dense: {
            type: Boolean,
            default: false
          },
          readonly: {
            type: Boolean,
            default: false
          }
        },
        data() {
          return {
            contentMeta
          };
        }
      };
    }
  },
  watch: {
    value: {
      handler() {
        Object.keys(this.$props.value).forEach(key => {
          this.content[key] = this.$props.value[key];
        });
      },
      immediate: true,
      deep: true
    },
    content: {
      handler() {
        this.$emit("input", this.content);
      },
      deep: true
    }
  }
};
</script>