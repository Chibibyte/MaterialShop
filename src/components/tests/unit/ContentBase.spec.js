import { shallowMount } from '@vue/test-utils';
import ContentBase from '@/components/ContentBase.vue';

let stubs = [
    'v-card',
    'v-container',
    'v-form'
]

let wrapper;
beforeEach(() => {
    wrapper = shallowMount(ContentBase, { stubs });
})

describe('Component', () => {
    test('is instantiated', () => {
        expect(wrapper.isVueInstance).toBeTruthy();
    })

    test('default snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    })

    test('props: contentMeta', () => {
        expect(wrapper.vm.contentMeta).toBeUndefined();
    })

    test('props: value', () => {
        expect(wrapper.vm.value).toEqual({});
    })

    test('props: disabled', () => {
        expect(wrapper.vm.disabled).toBeFalsy();
    })

    test('props: readonly', () => {
        expect(wrapper.vm.readonly).toBeFalsy();
    })

    test('props: dense', () => {
        expect(wrapper.vm.dense).toBeFalsy();
    })

    test('data: content', () => {
        expect(wrapper.vm.content).toEqual(wrapper.vm.value);
    })
})

// watch
describe('watch', () => {
    test('value', async () => {
        await wrapper.setProps({ value: { fieldA: 'contentA', fieldB: 'contentB' } });
        expect(wrapper.vm.content).toEqual(wrapper.vm.value);
    })

    test('content', async () => {
        wrapper.vm.content = { fieldA: 'contentA2', fieldB: 'contentB2' };
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted().input[0][0]).toEqual(wrapper.vm.content);
    })
})