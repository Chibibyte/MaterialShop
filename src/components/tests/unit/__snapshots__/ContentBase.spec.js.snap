// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component default snapshot 1`] = `
<v-card-stub>
  <v-container-stub
    class="white pb-0"
  >
    <v-form-stub
      class="white"
    />
     
  </v-container-stub>
</v-card-stub>
`;
